### PR TITLE
VariantContext to BQ row - Part2 (BQRow generator and pipeline runner)

### PR DESCRIPTION
**Pull request by BoscoSuen**
This part is the second part for the divided feature PR: VariantContext to BQ row.  
This part is based on the [first part](https://github.com/googleinterns/gcp-variant-transforms-java/pull/26) which provided functionalities to convert field value to the required format, the part2 provided the BQRowGenerator service which provides functionality to write ```VariantContext``` to BQ ```TableRow``` and set all fields. This part should be merged after the Part1
## Implementation
- **Pipeline Runner**: creating a Service interface **BigQueryRowGenerator.java**(and also its implementation class) and use the getRows() method in the interface to generate a **PCollection\<TableRow\>** and write the output to a JSON style string text file for demo test in the pipelineRunner. The header information comes from VcfToBqContext.
- **Generating BigQuery Row from VaraintContext(single VCF record)**: Creating a TableRow as a base row, and set all the field values from VariantContext, the tableRow should contain some nested rows for ALT fields and Calls(genotypes). More detailed notes are provided in the implementation. 

## Testing plan
As the unit test for all the Variant to BQ functionalities are provided in the part1(VariantToBqUtils), this part we will focus on the integration test to get record from VCF file and write to BQ table.
- **Integration test**: Read a sample VCF file, inject the **vcfParser**, **bigQueryRowGenerator**, test the tranformation from VCF file to VCFHeader, VCF record -> VariantContext -> Bigquery row. Test the row elements. Some corner cases testing are included.
- **Usage**: ```./gradlew test```

### CODE REVIEW

File: src/main/java/com/google/gcp_variant_transforms/options/VcfToBqOptions.java [hunk 1]
```diff
@@ -21,6 +22,16 @@
...
   void setInputFile(String value);
 
+  @Description("Flag for allowing malformed records")
+  @Default.Boolean(true)
+  public Boolean getAllowMalformedRecords();
+  void setAllowMalformedRecords(Boolean value);
+
+  @Description("Path of the file of malformed VCF records")
+  @Default.String("")
+  public String getMalformedRecordsMessage();
```
Comment (tneymanov) [inline src/main/java/com/google/gcp_variant_transforms/options/VcfToBqOptions.java:32]: MalformedRecordsReportPath maybe?

File: README.md [hunk 2]
```diff
@@ -29,5 +31,7 @@ OUTPUT=../output/report
   --region=${GOOGLE_CLOUD_REGION} \
   --tempLocation=${TEMP_LOCATION} \
   --inputFile=${INPUT_FILE} \
-  --output=${OUTPUT}"
+  --output=${OUTPUT} \
+  --allowMalformedRecords=${ALLOW_MALFORMED_RECORDS} \
+  --malformedRecordsMessage=${MALFORMEDRECORDSMESSAGE}"
```
Comment (tneymanov) [inline README.md:36]: Under underscores to MLAFORMEDRECORDSMESSAGE (or maybe better MALFORMED_RECORDS_REPORT_PATH)

File: src/main/java/com/google/gcp_variant_transforms/options/VcfToBqContext.java [hunk 3]
```diff
@@ -33,6 +37,9 @@ protected void validateFlags() throws IOException {
     if (this.output == null) {
       throw new IOException("No value for --output flag provided.");
     }
+    if (allowMalformedRecords && malformedRecordsMessage.isEmpty()) {
+      throw new IOException("Malformed records allowed but file path to malformed records not specified.");
```
Comment (tneymanov) [inline src/main/java/com/google/gcp_variant_transforms/options/VcfToBqContext.java:41]: Line above 100 chars, please look over other lines.

File: src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  @Inject public VcfParser vcfParser;
+
+  @Inject public BigQueryRowGenerator bigQueryRowGenerator;
+
+  /**
+   * Records             in the test VCF file:
+   * #CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	NA00001	NA00002	NA00003
+   * 20	14370	rs6054257	G	A	29	PASS	NS=3;DP=14;AF=0.5;DB;H2	GT:GQ:DP:HQ	0|0:48:1:51,51
+   * 1|0:48:8:51,51	1/1:43:5:.,.
+   * 20	17330	.	T	A	3	q10	NS=3;DP=11;AF=0.017	GT:GQ:DP:HQ	0|0:49:3:58,50	0|1:3:5:65,3	0/0:41:3
```
Comment (tneymanov) [inline src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java:74]: I guess lets ignore the line limit for this section...

File: src/test/resources/vcf_files_small_valid-4.0.vcf [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
+##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
+##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth">
+##FORMAT=<ID=HQ,Number=2,Type=Integer,Description="Haplotype Quality">
+#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	NA00001	NA00002	NA00003
+20	14370	rs6054257	G	A	29	PASS	NS=3;DP=14;AF=0.5;DB;H2	GT:GQ:DP:HQ	0|0:48:1:51,51	1|0:48:8:51,51	1/1:43:5:.,.
+20	17330	.	T	.	.	.	NS=3;DP=11;AF=0.017	GT:GQ:DP:HQ	./.:49:3:58,50	./.:3:5:65,3	./.:41:3
+20	1110696	rs6040355	A	G,T	67	PASS	NS=2;DP=10;AF=0.333,0.667;AA=T;DB	GT:GQ:DP:HQ	1|2:21:6:23,27	2|1:2:0:18,2	2/2:35:4
+20	1230237	.	T	.	47	PASS	NS=3;DP=13;AA=T	GT:GQ:DP:HQ	0|0:54:7:56,60	0|0:48:4:51,51	0/0:61:2
+19	1234567	microsat1	GTCT	G,GTACT	50	PASS	NS=3;DP=9;AA=G	GT:GQ:DP	0/1:35:4	0/2:17:2	1/1:40:3
```
Comment (tneymanov) [inline src/test/resources/vcf_files_small_valid-4.0.vcf:23]: Add new line.

File: src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java [hunk 6]
```diff
@@ -2,49 +2,74 @@
...
+    PCollection<VariantContext> variantContextPCollection = pipeline
+        .apply(TextIO.read().from(context.getInputFile()))
+        .apply(Filter.by((String inputLine) -> !inputLine.startsWith("#")))
+        .apply(ParDo.of(new ConvertLineToVariantFn(vcfParser, context.getHeaderLines())));
+    PCollectionTuple tableRowTuple = variantContextPCollection
+        .apply("VariantContextToBQRow",
+            ParDo.of(new ConvertVariantToRowFn(bigQueryRowGenerator,
+                context.getVCFHeader(), context.getAllowMalformedRecords(),
+                    VALID_VARIANT_TO_BQ_RECORD_TAG, MALFORMED_RECORD_ERROR_MESSAGE_TAG))
+                .withOutputTags(VALID_VARIANT_TO_BQ_RECORD_TAG, TupleTagList.of(MALFORMED_RECORD_ERROR_MESSAGE_TAG)));
```
Comment (tneymanov) [inline src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java:58]: Why is second one provided as list, while the first tuple tag is a single element? Is that the requirement?

Also, 100 chars.

Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java:58]: Yes. It's main tag + list of additional tags
